### DISCUSSION TITLE
Input-field don't lose focus when click outside input (inside modal with focus-trap)

### DISCUSSION THREAD

**Discussion by StensenE** [focus-trap/focus-trap#1171] [Q&A]
When I target a input field with f.ex type="search" and start typing and then click away, the input still have focus. The focus does not move along or beeing removed if no focus element is selected. The behaviour is fine with tabs. Have I missed something in the documentation? I've tried with allowOutsideClick: true, but that didn't work.  Right now I had to make a hack to make it work.

**Comment by stefcameron** [comment]
👋 @StensenE So you have an input element _inside_ a focus-trap, and the trap is _inside_ some type of  element active as a modal window. You type in the input and then click "away" -- "away" where?

Are you clicking _inside_ the modal, but outside the input? Or are you clicking _outside_ the modal completely?

If you're clicking _inside_ the modal, are there other tabbable or focusable elements _inside_ of it (i.e. siblings to the input)?

`allowOutsideClick: true` allows you to click on any element _outside the trap_ (presumably, in your case, the trap is the modal, so outside the modal) and having that element still register a click. Without this option set (it defaults to `false`), _nothing_ outside the trap will get a mouse event.

As for clicking _inside_ the trap, as long as the trap is active, it will strive to keep the focus inside of itself but will do nothing as long as the click is inside of itself. If you click on the "background" of your modal (i.e. the element that is the trap's container, let's say it just a `div`), focus-trap will do nothing (because it's inside the trap) and the browser will _remove_ focus from the input. Technically, at this point, `document.activeElement` will be the `<body>` and there's no reliable/efficient way to know that this has happened (I've tried). It's like being a strange state thanks to the browser.

And from this strange interim state, if all you do is press the TAB key, the browser will re-focus the input (not focus-trap, yet focus-trap will detect the focus but do nothing because it's still inside the trap) so you'll be back to where you were when you clicked outside the input, but inside the trap.

Finally, focus-trap is never trying to compete with the browser's focus behavior __unless__:
- focus departs from an edge node (e.g. you press the TAB key on the last element in tab order inside a configured trap container, in which case it will put focus on the next element in container order which will go against what the browser wants to do, hence the "trap");
- focus escapes the trap (in which case it will pull it back in to the most recently focused node); or
- you click on a focusable element outside the trap (in which case it will prevent that element from getting focus in order to keep focus inside the trap).

Outside of these conditions (and perhaps some corner cases I'm forgetting), focus-trap is letting the browser do its thing per normal DOM focus rules.

If you're wanting the trap to be __deactivated__ when you click outside the __trap__ (meaning you'd have to click somewhere outside your modal element), then the option you want to set is `clickOutsideDeactivates: true`

**Comment by StensenE** [comment]
@stefcameron This is only related to clicks inside modal (not tab, that works fine) 

When I click away from an input-element, but still inside the modal the focus from the first input won't blur. The issue seems to be related when I click from one input-elemnt to antoher inside modal. Like the case i'm facing are when users make a search and then click on checkbox for filtering the focus is still on search and the checkbox cannot be clicked unless I click back to search and then on checkbox again.

**Comment by stefcameron** [comment]
Interesting. So you have a trap, and it contains 2 inputs: 1 search and 1 checkbox. You can click into the search and type, but then if you want to check the box, you cannot unless you click it (which does nothing), then click back on the search, and then click back on the box (at which point it receives focus and gets checked). Yet you are able to tab from the search to the checkbox and press spacebar/enter to check it without issue?

I'd suggest the `allowOutsideClick: (element): boolean => yes/no` option but given both inputs are in the trap and the trap is active, there's no reason why both wouldn't be clickable/tabbable.

**Comment by StensenE** [comment]
Yes, when I start typing in search (input) the focus remain inside input, and it doens't matter where I click it seems. Without focus-trap everything works fine, so I made an hack for this particular case. allowOutsideClick don't seems to work with this case.

**Comment by stefcameron** [comment]
What does your "search" input look like? Is it a custom component, or just a plain HTML `<input type="text">` element? There must be something conflicting with what focus-trap is trying to do, although since both elements are inside the trap, focus-trap should basically be doing _nothing_ special other than checking that the click/focus/tab is happening inside its configured trap container.

Since you say you can tab to the next element but can't click to it, it would be interesting to know what focus-trap is doing in its [checkClick() handler](https://github.com/focus-trap/focus-trap/blob/master/index.js#L784), as well as its [checkFocusIn() handler](https://github.com/focus-trap/focus-trap/blob/master/index.js#L648).
